### PR TITLE
Problem: in zgossip, monitors are indexed by their pipe address and this causes trouble to zhashx

### PR DESCRIPTION
**Pull request by stvales**
Solution: index by string generated from monitor pipe address instead

### CODE REVIEW

**Comment by bluca** [comment]
The build is broken in several cases, and the zgossip test segfaults

**Comment by stvales** [comment]
"make test" is OK on macos and a fresh Centos install, with with drafts enabled or not on czmq and libzmq, and using the lastest libzmq master commit. I'll try tomorrow on Windows.

**Comment by bluca** [comment]
still crashing on osx

File: src/zgossip.c [hunk 1]
```diff
@@ -270,24 +270,26 @@ server_connect (server_t *self, const char *endpoint)
             zsock_destroy (&remote);
             return;
         }
-        
+#if defined (ZMQ_EVENT_HANDSHAKE_SUCCEEDED)
```
Comment (bluca) [inline src/zgossip.c:273]: this breaks existing stable API functionality - a fully working fallback is required

Comment (bluca) [reply src/zgossip.c:273]: or maybe not - was this existing functionality?

Comment (stvales) [reply src/zgossip.c:273]: The test on the definition of ZMQ_EVENT_HANDSHAKE_SUCCEEDED is supposed to be the fallback. If not defined, I do not create and store the zmonitors which rely on this event
This pull request does not touch the API. It applies a very specific bug correction in the server_connect and monitor_handler  functions of zgossip which already exist (I added monitor_handler with another pull request a few days ago).

Comment (stvales) [reply src/zgossip.c:273]: The attached script tests my czmq repo (and current pull request) with the latest masters of libzmq and libsodium on all four combinations of draft/non-draft API for libzmq and czmq. It passes perfectly on fresh macos 10.15 and CentOS 8 boxes. Same tests pass on a fresh Windows box using visual studio 2017.

Is there a way to restart tests from a fresh config  ?

-------------
git clone --depth 1 -b stable https://github.com/jedisct1/libsodium.git
cd libsodium
./autogen.sh && ./configure && make -j8
sudo make -j8 install
cd ..

git clone git://github.com/zeromq/libzmq.git
cd libzmq
mkdir build
mkdir build-draft
cd build
cmake .. -DCMAKE_BUILD_TYPE=Release -DWITH_LIBSODIUM=ON -DENABLE_DRAFTS=OFF
make -j8
sudo make install
cd ../build-draft
cmake .. -DCMAKE_BUILD_TYPE=Release -DWITH_LIBSODIUM=ON -DENABLE_DRAFTS=ON
make -j8
sudo ldconfig
cd ../..

git clone https://github.com/stvales/czmq.git
cd czmq
mkdir build
mkdir build-draft
cd build
cmake .. -DCMAKE_BUILD_TYPE=Release -DENABLE_DRAFTS=OFF
make -j8
make test
cd ../build-draft
cmake .. -DCMAKE_BUILD_TYPE=Release -DENABLE_DRAFTS=ON
make -j8
make test
cd ..

#switch to libzmq with drafts
cd ../libzmq/build-draft
sudo make install
cd ../../czmq/build
make test
cd ../build-draft
make  test

**Comment by stvales** [comment]
Closing to move to a brand new build pipeline.

**Pull request closed without merging**